design/ceph/osd-migration.md

```diff
@@ -0,0 +1,215 @@
+  1. List out all OSDs where `osd-store:<osd store type>` does not match with `spec.storage.store.type`.
+  2. If the PG status is not `active+clean`, don't replace any OSD.
+  3. If the PG status is `active+clean` but a previous OSD replacement is not completed, then don't replace any new OSD.
+  4. If the PG status is `active+clean` and no replacement is in progress, then select an OSD to be replaced from the list.
+  5. Delete the OSD deployment.
+  6. Create an OSD prepare job with an environment variable of OSD ID to be replaced.
+  7. OSD prepare pod will destroy the OSD and prepare it again using the same OSD ID. Refer `Destroy OSD` section below for details.
+  8. Once the destroyed OSD pod is added back, delete the configmap.
+  9. If there is any error during the OSD migration, then preserve the OSD being replaced in the configmap for next reconcile.
+  10. Reconcile the operator and perform as same steps until all the OSDs have migrated to the new backend store.
+
+#### Destroy OSD
+
+OSD prepare pod job will perform OSD deletion using following steps.
+- Check for OSD ID environment variable of the OSD to be destroyed.
+- Use `ceph volume list` to fetch the OSD path.
+- Destroy the OSD using following command:
+
+    ```
+    ceph osd destroy <OSD_ID> --yes-i-really-mean-it
+    ```
+
+- Wipe the OSD disk. This removes all the data on the device.
+- Prepare the OSD with the new store type by using the same OSD ID. This is done by passing the OSD ID as `--osd-id` flag to `ceph-volume` command.
+
+    ```
+    ceph-volume lvm prepare --osd-id <OSD_ID> --data <OSD_PATH>
+    ```
+
+#### Manage OSD migration status
+
+- OSD migration status will managed by `OSDReplaceConfig` API. This will be stored as the config map. For example:
+``` yaml
+apiVersion: v1
+data:
+  config: '{"OSDInprogress":{"id":0,"path":"/mnt/set1-data-045k6q","node":"minikube-m02"}'
+kind: ConfigMap
+metadata:
+  creationTimestamp: "2023-07-18T14:42:29Z"
+  name: osd-replace-config
+  namespace: rook-ceph
+  ownerReferences:
+  - apiVersion: ceph.rook.io/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: CephCluster
+    name: rook-ceph
+    uid: e634abc3-39fa-4dd0-a015-753030e64f95
+  resourceVersion: "3580"
+  uid: 7b7d35d4-7807-4286-b185-3f8562cb3cbe
+```
+- This config map will have the `ownerReference` of cephCluster resource.
+
+##### Ceph Cluster Status:
+
+- Ceph cluster status would indicate the progress of the OSD migration.
+- It will show the store type along with the number of OSDs that are using that store type.
+- Ceph cluster status should also be set to `Progressing` while migration is ongoing.
+- The status `message` will have some useful information about OSDs that are pending migration.
+  ``` yaml
+  status:
+   ceph:
+      osd:   # new status setting indicating the progress of OSD migration
+       storeType:
+        bluestore: 3
+        bluestore-rdr: 5
+      capacity:
+      bytesAvailable: 19212857344
```

Comment:
nit: I'd suggest removing the example status from this section that isn't directly related to the new osd status, so the reader can focus on what is new in the design

Reply:
removed the example status. Only keeping new OSD status.